cpp/include/raft/spatial/knn/detail/ivf_pq_build.cuh

```diff
@@ -108,55 +118,68 @@ struct bitfield_view_t {
   NB: label type is uint32_t although it can only contain values up to `1 << pq_bits`.
       We keep it this way to not force one more overload for kmeans::predict.
  */
-template <uint32_t PqBits>
-HDI void ivfpq_encode_core(uint32_t n_rows, uint32_t pq_dim, const uint32_t* label, uint8_t* output)
+template <uint32_t PqBits, size_t VecLen>
+__device__ void ivfpq_encode_core(uint32_t n_rows,
+                                  uint32_t pq_dim,
+                                  const uint32_t* label,
+                                  uint8_t* output)
 {
-  bitfield_view_t<PqBits> out{output};
-  for (uint32_t j = 0; j < pq_dim; j++, label += n_rows) {
-    out[j] = static_cast<uint8_t>(*label);
+  constexpr uint32_t kChunkSize = (VecLen * 8u) / PqBits;
+  TxN_t<uint8_t, VecLen> vec;
+  for (uint32_t j = 0; j < pq_dim;) {
+    vec.fill(0);
+    bitfield_view_t<PqBits> out{vec.val.data};
+#pragma unroll
+    for (uint32_t k = 0; k < kChunkSize && j < pq_dim; k++, j++, label += n_rows) {
+      out[k] = static_cast<uint8_t>(*label);
+    }
+    vec.store(output, 0);
+    output += VecLen;
   }
 }
 
 template <uint32_t BlockDim, uint32_t PqBits>
-__launch_bounds__(BlockDim) __global__
-  void ivfpq_encode_kernel(uint32_t n_rows,
-                           uint32_t pq_dim,
-                           const uint32_t* label,  // [pq_dim, n_rows]
-                           uint8_t* output         // [n_rows, pq_dim]
-  )
+__launch_bounds__(BlockDim) __global__ void ivfpq_encode_kernel(
+  uint32_t pq_dim,
+  const uint32_t* label,                                   // [pq_dim, n_rows]
+  device_mdspan<uint8_t, pq_codes_exts, row_major> output  // [n_rows, pq_dim]
+)
 {
   uint32_t i = threadIdx.x + BlockDim * blockIdx.x;
-  if (i >= n_rows) return;
-  ivfpq_encode_core<PqBits>(n_rows, pq_dim, label + i, output + (pq_dim * PqBits / 8) * i);
+  if (i >= output.extent(0)) return;
+  ivfpq_encode_core<PqBits, output.static_extent(2)>(
+    output.extent(0),
+    pq_dim,
+    label + i,
+    output.data_handle() + output.extent(1) * output.extent(2) * i);
 }
 }  // namespace
 
-inline void ivfpq_encode(uint32_t n_rows,
-                         uint32_t pq_dim,
+inline void ivfpq_encode(uint32_t pq_dim,
```

Comment:
/** Compress the cluster labels into an encoding with `pq_bits` bits, and transform it into a form to facilitaty vectorized loads */